### PR TITLE
JAVA-2624: Expose a config option for the connect timeout

### PR DESCRIPTION
**Pull request by emerkle826**

### CODE REVIEW

**Comment by emerkle826** [comment]
I tried using Simulacron to setup a situation where the client can't create a channel to the server and get the Netty connect timeout to pop, but Simulacron rejects immediately (for connections). It seems you can allow a connection and delay or prevent protocol negotiations from completing, but you can't have it delay the channel connection. If you feel I should do more to come up with a better test, let me know.

**Comment by tomekl007** [comment]
> I tried using Simulacron to setup a situation where the client can't create a channel to the server and get the Netty connect timeout to pop, but Simulacron rejects immediately (for connections). It seems you can allow a connection and delay or prevent protocol negotiations from completing, but you can't have it delay the channel connection. If you feel I should do more to come up with a better test, let me know.

I wonder how hard it will be to introduce such functionality in simulacron. I imagine that it may be useful in the future to emulate connection problems and test the delay on the connection.

File: core/src/main/resources/reference.conf
Comment (tomekl007) [inline core/src/main/resources/reference.conf]: why `5 seconds`?

Comment (emerkle826) [reply core/src/main/resources/reference.conf]: 5 seconds was the default for the 3.x driver: https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/SocketOptions.java#L23-L29

Comment (tomekl007) [reply core/src/main/resources/reference.conf]: Shouldn't we increase that? In driver 3 timeouts were more fine-grain. In 4.x the request timeout is taking whole path of request into account - so I suspect it is higher. In such a case maybe connect timeout should be also higher?

Comment (emerkle826) [reply core/src/main/resources/reference.conf]: This is just the Netty channel connection timeout though. I don't think it affects or is affected by request processing. Just how long Netty will wait before giving up trying to establish an I/O connection to the server.

File: core/src/main/java/com/datastax/oss/driver/internal/core/channel/ChannelFactory.java
Comment (emerkle826) [inline core/src/main/java/com/datastax/oss/driver/internal/core/channel/ChannelFactory.java]: @olim7t Should I move this bit into `afterBootstrapInitialized()`?

File: core/src/main/resources/reference.conf
Comment (olim7t) [reply core/src/main/resources/reference.conf]: 👍 
This is unrelated to `basic.request.timeout`. It will only affect session initialization, or background reconnections. The same default as 3.x is fine.

File: core/src/main/java/com/datastax/oss/driver/internal/core/channel/ChannelFactory.java
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/channel/ChannelFactory.java]: Yes, let's move it there for consistency, that's where the other socket options are processed.

File: core/src/main/resources/reference.conf
Comment (olim7t) [inline core/src/main/resources/reference.conf]: Nit: could you move it at the beginning of the section, either before or after the other timeouts? It feels slightly more convenient to have all the connection timeouts in the same place.

**Review by olim7t — APPROVED** [review]
LGTM, provided that we move the change to `NettyOptions` as discussed inline.

As for adding connect timeout to Simulacron, it would be a nice-to-have but I don't think we need to hold this PR for it, it's a pretty straightforward change. Maybe you can create a follow-up issue [here](https://github.com/datastax/simulacron/issues).

**Review by tomekl007 — APPROVED** [review]

**Comment by emerkle826** [comment]
I'll rebase this and resolve the conflict. I've also opened up an enhancement issue for Simulacron here: https://github.com/datastax/simulacron/issues/95

**Pull request merged**
